File: 0516-longest-palindromic-subsequence/0516-longest-palindromic-subsequence.cpp
class Solution {
public:
    int LCS(int i1, int i2, string &s1, string &s2, vector<vector<int>> &dp){
        if(i1 == 0 || i2 == 0) return 0;

        if(dp[i1][i2] != -1) return dp[i1][i2];

        if(s1[i1-1] == s2[i2-1]){
            return dp[i1][i2] = 1 + LCS(i1 -1, i2 -1, s1, s2, dp);
        }

        return dp[i1][i2] = max(LCS(i1 -1, i2, s1, s2, dp), LCS(i1, i2-1, s1, s2, dp));
    }

    int longestPalindromeSubseq(string s) {
        string s2 = s;
        reverse(s2.begin(), s2.end());

        int n = s.size();
        int m = s2.size();

        vector<vector<int>> dp(n+1, vector<int>(m+1, -1));
        return LCS(n, m, s, s2, dp);
    }
};

File: 0516-longest-palindromic-subsequence/0516-longest-palindromic-subsequence.py
class Solution:
    def findLCS(self, i, j, str1, str2, dp):
        if i < 0 or j < 0:
            return 0

        if dp[i][j] != -1:
            return dp[i][j]
        
        if str1[i] == str2[j]:
            return 1 + self.findLCS(i-1, j-1, str1, str2, dp)

        dp[i][j] = max(self.findLCS(i-1, j, str1, str2, dp), self.findLCS(i, j-1, str1, str2, dp))
        return dp[i][j]

    def longestPalindromeSubseq(self, s: str) -> int:
        n = len(s)
        s1 = s
        s2 = s[::-1]

        dp = [[-1 for _ in range(n)] for _ in range(n)]

        lcs = self.findLCS(n-1,n-1,s1,s2,dp)
        return lcs